File: lession1/CSharpDynamic/VisitorUsage/ObjectStructure.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using VisitorElement;
using VisitorExtension;

namespace VisitorUsage
{
    public class ObjectStructure
    {
        private IList<object> _list = new List<object>();
        public void Add(object element)
        {
            _list.Add(element);
        }

        public void Accept(dynamic visitor)
        {
            foreach (dynamic element in _list)
            {
                Accept(element, visitor);
            }
        }

        private void Accept(Element1 element, Visitor1 visitor)
        {
            visitor.View(element);
        }
        private void Accept(Element1 element, Visitor2 visitor)
        {
            visitor.View(element);
        }
        private void Accept(Element2 element, Visitor1 visitor)
        {
            visitor.View(element);
        }
        private void Accept(Element2 element, Visitor2 visitor)
        {
            visitor.View(element);
        }
    }
}

File: lession1/JavaNoPattern/VisitorJavaNoPatternFramework/src/main/java/ObjectStructure.java
/**
 * Created by Administrator on 2016/12/1.
 */
public class ObjectStructure {
    public static void dispatch(Object element, Object visitor) {
        if (element instanceof Element1 && visitor instanceof Visitor1) {
            ((Visitor1) visitor).view((Element1) element);
        }
        if (element instanceof Element2 && visitor instanceof Visitor1) {
            ((Visitor1) visitor).view((Element2) element);
        }
        if (element instanceof Element1 && visitor instanceof Visitor2) {
            ((Visitor2) visitor).view((Element1) element);
        }
        if (element instanceof Element2 && visitor instanceof Visitor2) {
            ((Visitor2) visitor).view((Element2) element);
        }

        if (element instanceof Element1 && visitor instanceof Visitor3) {
            ((Visitor3) visitor).view((Element1) element);
        }
        if (element instanceof Element2 && visitor instanceof Visitor3) {
            ((Visitor3) visitor).view((Element2) element);
        }
    }
}
